### ISSUE TITLE
Separation of UI presentation and graph

### ISSUE THREAD

**Issue by space-nuko** [comfyanonymous/ComfyUI#389]
Alot of visual programming tools I've used have a presentation mode separate from the graph mode, this is to reduce clutter and "wire forest" problems introduced by adding lots of complex nodes. Some examples:

<details>
<summary>Reaktor</summary>

![reaktor-6-screen-b](https://user-images.githubusercontent.com/24979496/229808664-af845807-b9f9-4c82-a53d-1a03e8708bdb.jpg)
</details>

<details>
<summary>Max/MSP</summary>

![FXPALLHHLZRS5CO](https://user-images.githubusercontent.com/24979496/229809770-be047ff6-285f-4f30-8481-92ac249373f1.png)
![FMK9Z25HLZRS5CP](https://user-images.githubusercontent.com/24979496/229809786-011f35ff-4c53-432f-ab44-401505b2b429.png)
</details>

Would this paradigm work for comfyUI? It would be nice to have a "build your own webui" paradigm where you can hide the node graph completely and work only with the UI you've built, once you've perfected the layout of your nodes

This may increase the frontend's complexity depending on the limitations of pure JS and may be an unacceptable change, but I would like using comfyUI a lot more if it were somehow implemented

Related https://github.com/comfyanonymous/ComfyUI/issues/278#issuecomment-1485975122

**Comment by space-nuko** [comment]
Just noticed https://github.com/comfyanonymous/ComfyUI/issues/42. I would support separating the frontend/backend in an obvious way to support building this kind of alternative UI

**Comment by comfyanonymous** [comment]
Yes if there was a way to "compile" a node graph into an interface it would be a nice feature.

>I would support separating the frontend/backend in an obvious way to support building this kind of alternative UI

The backend and frontend are already clearly separated.

Here's a simple example for how to use the api: https://github.com/comfyanonymous/ComfyUI/blob/master/script_examples/basic_api_example.py

Here's an example of an alternative frontend: https://github.com/rvion/CushyStudio

**Comment by DivinoAG** [comment]
The way Blender handles that might be a good principle to follow. There it is possible to create a group of nodes that can be collapsed into a single node in the main node tree (which is also a good idea here, but another subject). But to allow passing parameters the inner nodes, there are very basic nodes called Group Input and Group Output, which only function is to create an interface for the external view of that group.

For example, on the node tree below you have some coordinate inputs added to the Group Input, and the Fac property is linked to the Group Output

![image](https://user-images.githubusercontent.com/5044664/229907406-7c825935-19c4-4bfb-938d-0ae6a7731fc2.png)

And that makes it so when you view the group from the parent node tree, you see a nice interface showing all the parameters selected:

![image](https://user-images.githubusercontent.com/5044664/229908115-e575394e-0217-400d-927f-69c7d1e444a6.png)

Now, on ComfyUI, you could have similar nodes that, when connected to some inputs, these are displayed in a sidepanel as fields one can edit values without having to find them in the node workflow. All this UI node needs is the ability to add, remove, rename, and reoder a list of fields, and connect them to certain inputs from which they will inherent all values and formatting from. They don't require any crazy UI, you can display them mostly one below the other, similar to the UI for Easy Diffusion:

![image](https://user-images.githubusercontent.com/5044664/229908763-f387a7c5-58d9-4df4-ab9a-f55e7315dea1.png)

**Comment by justanothernguyen** [comment]
IMO we just need portal nodes, i.e. connecting to this one node will also connect to the other node somewhere else. A "wireless" node, if you will. After which a "control panel" can be made using such nodes.

Right now with reroute nodes and a bit of diligent, lots of visual clutters can already be done away with

From this
![image](https://user-images.githubusercontent.com/1379769/230230617-fba9bd53-d009-4d32-b8ed-39cbfb8ad2ac.png)

to this
![image](https://user-images.githubusercontent.com/1379769/230230668-66f72dcc-25ba-4596-9b6d-b1cf27183518.png)

**Comment by space-nuko** [comment]
That's what Max/MSP's presentation mode does, it creates a separate version of your graph with the wires erased, and when you go back into edit mode it restores all the wires
![max-intro-presentation-on-off](https://user-images.githubusercontent.com/24979496/230271437-23bbfc25-e4b8-498d-8612-fa7fcbe4a1c0.gif)

**Comment by Davemane42** [comment]
liteGraph does have support for subgraph but it is janky at best and multiple people already tried to implement it with mixed results.

Also, the "portal" node already exist as a custom extension (set/get node) but it keep breaking people workflows so 6ou might want to wait a bit.

A way to pin nodes/primitive to a collapsable side panel could be an easy way to display/change value on a big graph

**Comment by space-nuko** [comment]
I've been thinking also, litegraph is mostly canvas so it's not as good for accessibility as DOM elements. Plus with no way to anchor the node layout to the screen/prevent scrolling and zooming it would probably be tough to use as a standalone UI

**Comment by skyler14** [comment]
To add on the ideas here, it might make sense to make nodes and node groups all able to be activated/focused/etc. As ComfyUI gains prominence we'll see more tools trying to interface with SD beyond extensions and I think being able to then expand a graph node to a custom UI implementation may really have a lot of benefits.

**Comment by space-nuko** [comment]
Yeah I was thinking of being able to assign nodes "properties" like in a sidebar panel that affects things like grouping. So you could store the UI for each set of nodes in a separate tab, while they're all a part of the same graph. Or put them in an accordion.

Another thing, is being able to separate the individual widgets (sliders etc.) for each node panel and arrange them in a custom layout. I was thinking something like foobar2000's panel editing mode based on horizontal/vertical splitters, that translates to flexbox rows/columns. And defaulting to an alternative N-column layout for mobile users, both of which are saved to the workflow state

**Comment by skyler14** [comment]
As a foobar2000 user I know what you mean. 

I was thinking like this as far as how it could be an intuitive experience with nice extensibility

node groups:
-have a default (logical) interface which is just a list of all the inputs and outputs of the components in them
-you can define a (logical) interface that presents what inputs and outputs appear as edges on the node group
-when you double click/press edit/want to enter edit mode you expand your node group and are able to adjust the internal components in a draft to save

nodes:
-have a default (visual) interface which is just a list of all (the default to visible) options in them
-you can define a (visual) interface that is basically made to treat the backend stuff making up that node like an api, it defaults to expanding to show a list of everything (like some kind of map of all the variables as some kind of input field + formatting), but just as easily the variables can be fed into a script to create a more elaborate UI (picture like someone had an viewer node for checking the results of an intermediate upsampling step and they wanted to add a image compare slider).
-when you double click on an element either the default visual interface pops up or any custom interfaces you have loaded up/prepared/uploaded

**Comment by space-nuko** [comment]
Hmm, sounds interesting, I'm having some trouble visualizing what this would look like though. Could you provide some mockups if it's not too much trouble?

Your idea for a default interface sounds like the "mobile version" thing I had in mind, it's just a set of columns that can be collapsed into one single column on mobile or multiple on desktop depending on window size

**Comment by DivinoAG** [comment]
I think there are many different ways to approach this concept described by @skyler14. 

The Node Groups UI could work exact as my previous example from Blender. You select a number of nodes and create a new group from them (Ctrl+G, or right clicking and selecting Create Node Group from a menu, for example). Every group automatically have Group Inputs and Group Outputs nodes. You plug any cable onto the empty connector on the Group Input/Output, and it adds a new parameter to that group. Double click that input/output name to rename them to whatever for clarity. When you close that group, only the parameters on those Input/Output nodes get displayed in the new group node. I'm posting the image here again for context.

![image](https://user-images.githubusercontent.com/5044664/234123773-f9b3c488-441c-4da3-a614-7af083b82106.png)

As for the general node UI, the simplest way to achieve this I think would be to automatically display every parameter available to the user in a sidepanel, grouped by the node they belong to, and in the order those nodes are processed when you run the prompt. Then, give the user the option to hide the nodes/parameters they don't need to see for that workflow, so they end up with only the list of parameters they need to change in order to make their images.

I made a quick mockup of how I see that working, using my image upscaling workflow as an example:

**Side panel showing all the editable parameters**
![image](https://user-images.githubusercontent.com/5044664/234134792-b10055cc-bb9a-49a2-83ec-dbc1202d66de.png)

**Editing the visible parameters**
![image](https://user-images.githubusercontent.com/5044664/234134839-92fd1387-50c2-43a8-af43-a56f1d1c4311.png)

**Side panel showing only the selected parameters**
![image](https://user-images.githubusercontent.com/5044664/234134877-9d3ffb6f-f6fd-43be-848e-75e6ab6f34e1.png)

Now, there is more that could done with this side panel, but I think this is an MVP that could get the ball rolling, and isn't terribly complicated to both develop, and to improve usability for users. I also didn't do anything too crazy in design here, mostly it's the same look we know and are used to. It could probably be improved, but that isn't truly necessary to make this work.

**Comment by space-nuko** [comment]
@DivinoAG I've actually done this already with my alternative UI, except not in a sidebar but split into top and bottom. Nothing stopping it from being moved there in my mind.

https://github.com/space-nuko/ComfyBox

**Comment by 0xdevalias** [comment]
Just stumbled onto this other issue about subgraphs while trying to find this one again, seems to be a lot of crossover potential in ideas and implementations being discussed between them:

- https://github.com/Comfy-Org/rfcs/pull/15
  - https://github.com/Comfy-Org/rfcs/discussions/17
  - https://github.com/Comfy-Org/rfcs/blob/main/rfcs/0005-subgraph.md
- https://github.com/comfyanonymous/ComfyUI/issues/42
- https://github.com/comfyanonymous/ComfyUI/issues/497
- https://github.com/comfyanonymous/ComfyUI/issues/669
- https://github.com/comfyanonymous/ComfyUI/pull/724
- https://github.com/comfyanonymous/ComfyUI/pull/931
- https://github.com/comfyanonymous/ComfyUI/issues/1132
- https://github.com/comfyanonymous/ComfyUI/pull/1310
- https://github.com/comfyanonymous/ComfyUI/issues/1355
- https://github.com/comfyanonymous/ComfyUI/issues/1412
- https://github.com/comfyanonymous/ComfyUI/pull/1776
- https://github.com/comfyanonymous/ComfyUI/issues/2122
- https://github.com/comfyanonymous/ComfyUI/issues/2483
- https://github.com/Comfy-Org/ComfyUI_frontend/issues/1077
- https://github.com/ltdrdata/ComfyUI-Workflow-Component
  - > This is a side project to experiment with using workflows as components.

---

Also wanted to include this idea/note I posted on the other issue here; RE: litegraph bugs/limitations and a potentially better alternative:

> > I think it will require several changes in litegraph to fix bugs. 
> 
> > These changes are in my fork of litegraph which isn't API compatible with upstream litegraph however as the pace of development is too significant
> 
> This is probably a little off topic for this issue, and I suspect that it would be a rather major / potentially breaking change to do so.. but I'm curious if there's a reason that litegraph is used instead of something like xyflow / reactflow / etc? I haven't deeply evaluated both projects yet, but at least in my initial cursory exploration, xyflow seems like it's capable of everything liteflow does, but in a more standard are modern developed/maintained way.
> 
> -   <https://github.com/xyflow/xyflow>
>     -   > React Flow | Svelte Flow - Powerful open source libraries for building node-based UIs with React ([https://reactflow.dev](https://reactflow.dev/)) or Svelte ([https://svelteflow.dev](https://svelteflow.dev/)). Ready out-of-the-box and infinitely customizable.
>     -   <https://www.xyflow.com/>
>         -   > Powerful open source libraries for building node-based UIs with React or Svelte. Ready out-of-the-box and infinitely customizable
>     -   [https://reactflow.dev](https://reactflow.dev/)
>         -   > Wire Your Ideas with React Flow A customizable React component for building node-based editors and interactive diagrams
>     -   [https://svelteflow.dev](https://svelteflow.dev/)
>         -   > Wire Your Ideas with Svelte Flow A customizable Svelte component for building node-based editors and interactive diagrams by the creators of React Flow
> 
> _Originally posted by @0xdevalias in https://github.com/comfyanonymous/ComfyUI/issues/669#issuecomment-1833179664_

**Comment by ZapAndersson** [comment]
The "presentation mode" of Max/MSP is much much simpler, though. It's a mode of a while graph, and when it is on, the **only** differences are

- only nodes flagged for "presentation mode" are displayed
- nodes more to an alternate "presentation mode" position

That's it! In theory, this should be pretty simple to implement, no?

**Comment by 0xdevalias** [comment]
Some recent updates on related issues:

> https://blog.comfy.org/p/subgraphs-are-coming-to-comfyui 
>
> _Originally posted by @phazei in https://github.com/comfyanonymous/ComfyUI/issues/669#issuecomment-3134677826_

> - https://blog.comfy.org/p/subgraph-official-release
>   - > Subgraph Official Release! Making Complex Workflows Clean and Efficient
>   - > ComfyUI Subgraph: Package complex workflows into reusable nodes. Transform chaos into simple, modular components. 
>
> _Originally posted by @0xdevalias in https://github.com/comfyanonymous/ComfyUI/issues/669#issuecomment-3194156293_

> - https://blog.comfy.org/p/comfyui-035-frontend-updates
>   - > ComfyUI 0.3.51: Subgraph, New Manager UI, Mini Map and More
>   - > Subgraph is officially available in stable releases, and it now supports unpacking a subgraph back into its original nodes on the main graph.
>   - > And the subgraph feature is still evolving. Upcoming improvements include:
>     > 
>     > - Publishing subgraphs as reusable nodes
>     > - Synchronizing updates across linked subgraph instances
>     > - Automatically generating subgraph widgets 
> 
> _Originally posted by @0xdevalias in https://github.com/comfyanonymous/ComfyUI/issues/669#issuecomment-3212888917_

---

The one that I thought sounded particularly interesting/relevant was the 'automatically generating subgraph widgets' mentioned in that last one as an 'upcoming improvement'.

I haven't searched too deeply to find more info about it; but at face value it definitely felt like it could relate to more 'user friendly' user-facing UI elements backed by a subgraph.

**Comment by ghostsquad** [comment]
I found that having groups of nodes (or subgraphs), and using rgthree context to pass information along to be very powerful. Then I mute whole groups. The problem remains though that, the group is still visible, it's still taking up space. My workflow UI is bigger than what fits on the screen, but that's not usually the case if these groups (or now it would be subgraphs) were hidden when muted.
